### PR TITLE
Added the equivalent of np.who()

### PR DESCRIPTION
**Pull request by OsamaNabih**
This is my first commit to CuPy, this is a simple function. The code is almost identical to NumPy's. I know it's nothing much, I'm just getting familiar with the process and CuPy itself. :)
I wasn't sure how this function should be categorized, so I made a new file to misc folder.
Please let me know of any adjustments required.

### CODE REVIEW

File: cupy/misc/who.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    >>> cp.who(d)
+    Name            Shape            Bytes            Type
+    ===========================================================
+    x               2                16               float64
+    y               3                24               float64
+    Upper bound on total bytes  =       40
+    """
+
+    if vardict is None:
+        frame = sys._getframe().f_back
```
Comment (emcastillo) [inline cupy/misc/who.py]: why not `globals()`?

File: cupy/misc/who.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import sys
+import cupy as cp
+
+
+def who(vardict=None):
+
+    """
+    Print the CuPy arrays in the given dictionary.
```
Comment (emcastillo) [inline cupy/misc/who.py]: docstring does not align with CuPy format

**Comment by emcastillo** [comment]
Add tests for valid and invalid cases and also fix travis errors please.

File: cupy/misc/who.py (hunk 1 above)
Comment (OsamaNabih) [reply cupy/misc/who.py]: Wouldn't that return the stack the function is in, not the stack it was called from?

File: cupy/misc/who.py (hunk 2 above)
Comment (OsamaNabih) [reply cupy/misc/who.py]: Could you please point me to somewhere that explains the difference? I'm not sure what the Google style is

File: cupy/misc/who.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            print(prval + "\n" + "="*(len(prval)+5) + "\n")
+
+        for k in range(len(sta)):
+            val = sta[k]
+            print("%s %s %s %s %s %s %s" % (val[0], ' '*(sp1-len(val[0])+4),
+                                            val[1], ' '*(sp2-len(val[1])+5),
+                                            val[2], ' '*(sp3-len(val[2])+5),
+                                            val[3]))
+        print("\nUpper bound on total bytes  =       %d" % totalbytes)
+    return
```
[the remark this answers is not in the document]
Comment (OsamaNabih) [reply cupy/misc/who.py]: Okay

File: cupy/misc/who.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import sys
+import cupy as cp
```
[the remark this answers is not in the document]
Comment (OsamaNabih) [reply cupy/misc/who.py]: Alright, thanks for letting me know

File: cupy/misc/who.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    >>> d = {'x': cp.arange(2.0), 'y': cp.arange(3.0), 'txt': 'Some str',
+    ... 'idx':5}
+    >>> cp.who(d)
+    Name            Shape            Bytes            Type
+    ===========================================================
+    x               2                16               float64
+    y               3                24               float64
+    Upper bound on total bytes  =       40
+    """
+
```
[the remark this answers is not in the document]
Comment (OsamaNabih) [reply cupy/misc/who.py]: Alright, done

**Comment by OsamaNabih** [comment]
> 
> 
> Add tests for valid and invalid cases and also fix travis errors please.

Could you add suggestions for invalid cases? I'm not sure what that would be.

File: cupy/misc/who.py (hunk 2 above)
Comment (emcastillo) [reply cupy/misc/who.py]: You can check other routines in the code.

File: cupy/misc/who.py (hunk 1 above)
Comment (emcastillo) [reply cupy/misc/who.py]: Good point

**Comment by emcastillo** [comment]
> Could you add suggestions for invalid cases? I'm not sure what that would be.

Passing invalid parameters, a list instead of a dictionary and checking that it raises the corresponding error.

**Comment by emcastillo** [comment]
Also please fix travis errors:
https://travis-ci.org/github/cupy/cupy/builds/668283579?utm_source=github_status&utm_medium=notification

And add the function to the documentation in the corresponding doc/*.rst file

**Comment by emcastillo** [comment]
@OsamaNabih hi!
Sorry for getting back late.
Are you planning to work on this?

Some example of docstring is 

https://github.com/cupy/cupy/blob/master/cupy/math/rounding.py#L7-L19

Also for test invalid cases, I meant passing the `np.who` function values other than a dictionary.

Thanks for the contribution!

**Comment by kmaehashi** [comment]
Moved to #3361

**Pull request closed without merging**